tests/tool_use/conftest.py

```diff
@@ -0,0 +1,30 @@
+import pytest
+from huggingface_hub import snapshot_download
+
+from tests.utils import RemoteOpenAIServer
+
+from .util import ARGS, CONFIGS, ServerConfig
+
+
+# for each server config, download the model and return the config
+@pytest.fixture(scope="session", params=CONFIGS.keys())
+def server_config(request):
+    config = CONFIGS[request.param]
+    # download model and tokenizer using transformers
+    snapshot_download(config["model"])
+    yield CONFIGS[request.param]
+
+
+# run this for each server config
+@pytest.fixture(scope="session")
```

Comment:
That's the default behavior of pytest fixtures. It sets up a fixture, runs the tests, and then tears it down. For parameterized fixtures, it does each of those steps in sequence for each parameter unless you use a plugin to specify that you want them to run concurrently. 

The `server` and `client` fixtures are downstream of the `server_config` fixture because they request it in their parameters. `server_config` is parameterized, so the order (because of pytest's default behavior) is to get the first parameter value for the `server_config` fixture (`"hermes"` or `"mistral"`). It runs the `server_config` fixture for the value, then runs `server`, then runs `client`, and then calls all the tests, which all request the `client` fixture. Once tests finish, pytest cleans up the fixtures (in this case, it's handled in `RemoteOpenAIServer`) which is any code after `yield`. Then it uses `server_config` again with the next parameter value, and repeats the entire process. 

The issue that we ran into earlier was because the server took longer to start up than expected resulting in a timeout error that didn't result in the server shutting down as expected.

Hope that makes sense, happy to explain or discuss further if necessary :)

Reply:
I spent way too much time in the pytest docs trying to figure out how to do this correctly without code duplication ☠️